Mathlib/RepresentationTheory/Coinvariants.lean

```diff
@@ -0,0 +1,405 @@
+  rfl
+
+end TensorProduct
+end
+end Representation
+
+namespace Rep
+
+open CategoryTheory Representation
+
+variable {k G : Type u} [CommRing k] [Monoid G] {A B C : Rep k G} {n : ℕ}
+
+namespace Coinvariants
+
+/-- The linear map underlying a `G`-representation morphism `A ⟶ B`, where `B` has the trivial
+representation, factors through `A_G`. -/
+noncomputable abbrev lift [B.ρ.IsTrivial] (f : A ⟶ B) :
+    A.ρ.Coinvariants →ₗ[k] B :=
+  Representation.Coinvariants.lift _ f.hom.hom fun _ => by
+    ext
+    have := hom_comm_apply f
+    simp_all
+
+/-- A `G`-representation morphism `A ⟶ B` induces a linear map `A_G →ₗ[k] B_G`. -/
+noncomputable abbrev map (f : A ⟶ B) :
+    A.ρ.Coinvariants →ₗ[k] B.ρ.Coinvariants :=
+  Representation.Coinvariants.map _ _ f.hom.hom fun _ => ModuleCat.hom_ext_iff.1 (f.comm _)
+
+@[simp]
+theorem map_id (A : Rep k G) :
+    map (𝟙 A) = LinearMap.id := by
+  ext; rfl
+
+@[simp]
+theorem map_comp (f : A ⟶ B) (g : B ⟶ C) :
+    map (f ≫ g) = map g ∘ₗ map f := by
+  ext; rfl
+
+variable (k G)
+
+/-- The functor sending a representation to its coinvariants. -/
+@[simps obj map]
+noncomputable def functor : Rep k G ⥤ ModuleCat k where
+  obj A := ModuleCat.of k (A.ρ.Coinvariants)
+  map f := ModuleCat.ofHom (map f)
+
+variable {k G} in
+@[ext]
+lemma functor_hom_ext {M : ModuleCat k} {f g : (functor k G).obj A ⟶ M}
+    (hfg : f.hom ∘ₗ Coinvariants.mk A.ρ = g.hom ∘ₗ Coinvariants.mk A.ρ) :
+    f = g := ModuleCat.hom_ext <| Coinvariants.hom_ext hfg
+
+instance : (functor k G).Additive where
+  map_add := ModuleCat.hom_ext <| LinearMap.ext fun x => Quotient.inductionOn' x (fun _ => rfl)
+
+/-- The adjunction between the functor sending a representation to its coinvariants and the functor
+equipping a module with the trivial representation. -/
+noncomputable def adjunction : functor k G ⊣ trivialFunctor G :=
+  Adjunction.mkOfHomEquiv {
+    homEquiv X Y := {
+      toFun f := {
+        hom := ModuleCat.ofHom (f.hom ∘ₗ Coinvariants.mk X.ρ)
+        comm g := by
+          ext x
+          exact congr(f.hom $((Coinvariants.mk_eq_iff X.ρ).2
+            (X.ρ.mem_augmentationSubmodule_of_eq g x _ rfl))) }
+      invFun f := ModuleCat.ofHom (lift f)
+      left_inv _ := by ext; rfl
+      right_inv _ := Action.Hom.ext <| rfl }
+    homEquiv_naturality_left_symm _ _ := by ext; rfl
+    homEquiv_naturality_right := by intros; rfl }
+
+instance : (functor k G).PreservesZeroMorphisms where
+  map_zero _ _ := by ext; rfl
+
+instance : Limits.PreservesColimits (functor k G) :=
+  (adjunction k G).leftAdjoint_preservesColimits
+
+end Coinvariants
+
+variable (k G) in
+open MonoidalCategory ModuleCat.MonoidalCategory in
+/-- The functor sending `A, B` to `(A ⊗[k] B)_G`. This is naturally isomorphic to the functor
+sending `A, B` to `A ⊗[k[G]] B`, where we give `A` the `k[G]ᵐᵒᵖ`-module structure defined by
+`g • a := A.ρ g⁻¹ a`. -/
+@[simps]
+noncomputable def coinvariantsTensor : Rep k G ⥤ Rep k G ⥤ ModuleCat k where
```

Comment:
This should be using `Functor.postcompose\_2` applied to the "curried tensor" bifunctor, and should probably be an abbrev.